src/impl.jl

```diff
@@ -302,24 +302,69 @@ function auto_hash_equals_impl(__source__, struct_decl, fields, cache::Bool, has
         end))
     end
 
-    # Add the == function
-    equalty_impl = foldl(
-        (r, f) -> :($r && $isequal($getfield(a, $(QuoteNode(f))), $getfield(b, $(QuoteNode(f))))),
-        fields;
-        init = cache ? :(a._cached_hash == b._cached_hash) : true)
-    if struct_decl.args[1]
-        # mutable structs can efficiently be compared by reference
-        equalty_impl = :(a === b || $equalty_impl)
-    end
-    if isnothing(where_list) || !typearg
-        push!(result.args, esc(:(function $Base.:(==)(a::$type_name, b::$type_name)
-            $equalty_impl
-        end)))
-    else
-        # If requested, require the type arguments be the same for two instances to be equal
-        push!(result.args, esc(:(function $Base.:(==)(a::$full_type_name, b::$full_type_name) where {$(where_list...)}
-            $equalty_impl
-        end)))
+    # Add the `==` and `isequal` functions
+    for eq in (==, isequal)
+        if eq == isequal
+            equalty_impl = foldl(
+                (r, f) -> :($r && $eq($getfield(a, $(QuoteNode(f))), $getfield(b, $(QuoteNode(f))))),
+                fields;
+                init = cache ? :(a._cached_hash == b._cached_hash) : true)
+            if struct_decl.args[1]
+                # mutable structs can efficiently be compared by reference
+                # Note this optimization is only valid for `isequal`, e.g.
+                # a = [missing]
+                # a == a # missing
+                # isequal(a, a) # true
+                equalty_impl = :(a === b || $equalty_impl)
+            end
+        else
+            # Here we have a more complicated implementation in order to handle missings correctly.
+            # If any field comparison is false, we return false (even if some return missing).
+            # If no field comparisons are false, but one comparison missing, then we return missing.
+            # Otherwise we return true.
+            # (This matches the semantics of `==` for `Tuple`'s and `NamedTuple`'s.)
+
+            # Here we do some manual hygiene, since we will escape everything at the end
+            found_missing = gensym(:found_missing)
+            cmp = gensym(:cmp)
+            equalty_impl = quote
```

Comment:
I am not super familiar with macros, do you have a suggested alternate implementation?

Reply:
I tried switching to `:()`, but I think there's still line numbers getting in:
```julia
 @macroexpand @auto_hash_equals fields=(a,b) struct Foo
                           a
                           b
                           c
                       end
```
gives
```julia
    function (Base).:(==)(a::Foo, b::Foo)
        #= /Users/eph/AutoHashEquals.jl/src/impl.jl:348 =#
        #= /Users/eph/AutoHashEquals.jl/src/impl.jl:349 =#
        $(Expr(:(=), :found_missing, false, :(cmp = (isequal)((getfield)(a, :a), (getfield)(b, :a))), :(#= /Users/eph/AutoHashEquals.jl/src/impl.jl:337 =#), :(cmp === false && return false), :(#= /Users/eph/AutoHashEquals.jl/src/impl.jl:338 =#), :((ismissing)(cmp) && (found_missing = true)), :(cmp = (isequal)((getfield)(a, :b), (getfield)(b, :b))), :(#= /Users/eph/AutoHashEquals.jl/src/impl.jl:337 =#), :(cmp === false && return false), :(#= /Users/eph/AutoHashEquals.jl/src/impl.jl:338 =#), :((ismissing)(cmp) && (found_missing = true)), :((ifelse)(found_missing, missing, true))))
    end
```
This version also puts everything on 1 line which seems more confusing.

Reply:
I suggest to build the function expression nodes yourself, line by line, rather than using quotation at all.

But since the current implementation also produces these lines, don't worry about it.  I'll create an issue to deal with it later.